library/core/src/ptr/mod.rs

```diff
@@ -2388,4 +2407,105 @@ mod verify {
         let copy = unsafe { read_volatile(ptr) };
         assert_eq!(val, copy);
     }
+
+    fn check_align_offset<T>(p: *const T) {
+        let a = kani::any::<usize>();
+        unsafe { align_offset(p, a) };
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_zst() {
+        let p = kani::any::<usize>() as *const ();
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_u8() {
+        let p = kani::any::<usize>() as *const u8;
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_u16() {
+        let p = kani::any::<usize>() as *const u16;
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_u32() {
+        let p = kani::any::<usize>() as *const u32;
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_u64() {
+        let p = kani::any::<usize>() as *const u64;
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_u128() {
+        let p = kani::any::<usize>() as *const u128;
+        check_align_offset(p);
+    }
+
+    #[kani::proof_for_contract(align_offset)]
+    fn check_align_offset_4096() {
+        let p = kani::any::<usize>() as *const [u128; 64];
+        check_align_offset(p);
+    }
+
+    // This function lives inside align_offset, so it is not publicly accessible (hence this copy).
+    #[safety::requires(m.is_power_of_two())]
+    #[safety::requires(x < m)]
+    unsafe fn mod_inv_copy(x: usize, m: usize) -> usize {
```

Comment:
Is this copy for the purpose of calling the function from a harness? If so, we should be able to write the harness inside `align_offset` to avoid the copy.

For instance, this works:
```rust
fn foo(x: u32) -> u32 {
    fn internal(x: u32) -> u32 {
        x / 2
    }
    
    #[kani::proof]
    fn check_internal() {
        let x = kani::any();
        let y = internal(x);
        assert!(y <= x);
    }

    let y = internal(x);
    y / 2

}

#[kani::proof]
fn check_foo() {
    let x = kani::any();
    let y = foo(x);
    assert!(y <= x / 2);
}
```

```bash
$ kani internal_harness.rs  | grep harness
Checking harness check_foo...
Reading GOTO program from file /home/ubuntu/examples/internal_harness__RNvCs97HbwSwwU9k_16internal_harness9check_foo.out
         - Location: internal_harness.rs:14:5 in function foo
         - Location: internal_harness.rs:14:5 in function foo
         - Location: internal_harness.rs:22:18 in function check_foo
         - Location: internal_harness.rs:22:18 in function check_foo
         - Location: internal_harness.rs:22:5 in function check_foo
         - Location: internal_harness.rs:3:9 in function foo::internal
         - Location: internal_harness.rs:3:9 in function foo::internal
Checking harness foo::check_internal...
Reading GOTO program from file /home/ubuntu/examples/internal_harness__RNvNvCs97HbwSwwU9k_16internal_harness3foo14check_internal.out
         - Location: internal_harness.rs:10:9 in function foo::check_internal
         - Location: internal_harness.rs:3:9 in function foo::internal
         - Location: internal_harness.rs:3:9 in function foo::internal
Complete - 2 successfully verified harnesses, 0 failures, 2 total.
```

Reply:
I tried this, but I'm getting:

```
error: Failed to compile `core` due to an internal compiler error.: error: internal compiler error: Kani unexpectedly panicked at panicked at kani-compiler/src/codegen_cprover_gotoc/codegen/contract.rs:46:14:
                                called `Option::unwrap()` on a `None` value.
```

I did some debugging and got some information, but I didn't spot the bug. I'll open an issue with what I found.

Reply:
If we're okay merging it as-is pending the solution to this issue, I can add a TODO to get rid of the harness once the issue is fixed.

Reply:
Sounds good.